### ISSUE TITLE
Java agent missing trace-id/memory leak when making a rest call to Forgerock identity gateway stand alone version with http2 prior knowledge

### ISSUE THREAD

**Issue by muhardanz** [open-telemetry/opentelemetry-java-instrumentation#8796]
**Describe the bug**
When using open telemetry java agent with Forgerocks Identity gateway standalone version there is a memory leak when making requests using http2 prior knowledge.

This does not happen when making a http2 connection that steps up from http1 or requests that are simply made with http1

**Steps to reproduce**
run the identity gateway standalone app from Forgerock adding the java agent to the command line arguments (start up guide can be found here: https://backstage.forgerock.com/docs/ig/2023.4/getting-started/start-product.html)
Once the service is running (assuming it is running on port 8080) make a call to any route (or invalid route):
curl --location 'http://localhost:8080/invalidroute' -v --http2-prior-knowledge

**What did you expect to see?**
The following information should be seen in the logs with each log message
{
    "trace_id": "e4b3c387b779c29005d4fbdff5a30373",
    "trace_flags": "01",
    "span_id": "0362976de5f1670f"
}

**What did you see instead?**
No Trace ID printed in the logs and also Out Of Memory exception will occur if a lot of http2 prior knowledge calls are made

**What version are you using?**
v1.27.0

**Environment**
NAME="Red Hat Enterprise Linux"
VERSION="8.8 (Ootpa)"
ID="rhel"
ID_LIKE="fedora"
VERSION_ID="8.8"
PLATFORM_ID="platform:el8"
PRETTY_NAME="Red Hat Enterprise Linux 8.8 (Ootpa)"
Forgerock identity gateway version: 2023.4

**Comment by laurit** [comment]
I started Identity Gateway 2023.4.0 with `JAVA_OPTS="-Xmx256m -XX:+HeapDumpOnOutOfMemoryError -javaagent:opentelemetry-javaagent-1.27.0.jar" ./start.sh` and ran `while true; do curl --location 'http://localhost:8080/invalidroute' -v --http2-prior-knowledge; done` There was no oom, heap size remained small.

**Comment by muhardanz** [comment]
Thanks [laurit](https://github.com/laurit)
I will attempt the same test and also verify in the logs and then get back to you

**Comment by muhardanz** [comment]
I also cannot seem to generate the OOM using the latest version of the agent on my MAC using the same commands.
The original OOM was seen in a Kuberenetes cluster that was running for a significantly long time.

That said I can observe the buggy behaviour of trace-ids not being picked up for http2 prior knowledge:


**Starting IG so that you can view logs**

1. `export JAVA_OPTS="-javaagent:opentelemetry-javaagent-1.27.0.jar"`

2. This command is required to view otel behaviour in the IG log output
`export OTEL_TRACES_EXPORTER="logging"`

3. Start ig in the foreground to view logs
`./start.sh`

**Testing calls to IG in a seperate terminal**

1. Do the following step incase your system is setup with a proxy that may interfere with http2 prior knowledge
`unset {http,https,all}_proxy && unset {HTTP,HTTPS,ALL}_PROXY`

2. Run a curl test on an IG out of the box endpoint or an invalid route:
`curl --location 'http://localhost:8080/invalidroute' -v --header 'traceparent: 00-21a972e33b7d4ba59bacfa567653a5e1-055fa9a6e05a46bf-01' -v --http2-prior-knowledge`
At this point no log will be present for otel in the ig console logs

3. Run the same test without http2 prior knowledge
`curl --location 'http://localhost:8080/openig/metrics/prometheus' -v --header 'traceparent: 00-21a972e33b7d4ba59bacfa567653a5e1-055fa9a6e05a46bf-01'`
At this point the otel log will show with the trace-id on the ig console log:
21a972e33b7d4ba59bacfa567653a5e1


These tests show that the trace-id is not being picked up by the agent when http2 prior knowledge is used

**Comment by muhardanz** [comment]
**Further Testing**
First run the steps to run IG in the forground and unset the proxy for these tests

1. create a custom route in the path ~/.openig/config/routes/001-test-route.json
```
{
  "capture": "all",
  "name": "Test Route",
  "condition": "${find(request.uri.path,'^/test-route$')}",
  "handler": {
    "type": "Chain",
    "config": {
      "filters": [
        {
          "type": "UriPathRewriteFilter",
          "config": {
            "mappings": {
              "/test-route": "/openig/metrics/prometheus"
            }
          }
        }
      ],
      "handler": {
        "type": "ReverseProxyHandler"
      }
    }
  }
}
```
This route simply creates a reverse proxy to itself as a the downstream connection

2. Run a curl test on an IG custom route:
`curl --location 'http://localhost:8080/test-route' --header 'traceparent: 00-21a972e33b7d4ba59bacfa567653a5e1-055fa9a6e05a46bf-01' -v --http2-prior-knowledge`
At this point the downstream trace id will not be correct. It seems to generate a fresh one

3. Run the same test without http2 prior knowledge
`curl --location 'http://localhost:8080/test-route' --header 'traceparent: 00-21a972e33b7d4ba59bacfa567653a5e1-055fa9a6e05a46bf-01' -v`
At this point the otel log will show with the correct trace-id on the downstream request in ig console log:
21a972e33b7d4ba59bacfa567653a5e1

**Comment by laurit** [comment]
@muhardanz If you want us to look into the Out of Memory you'll have to provide a heap dump or instructions to reproduce or comparable evidence that would show that this is an issue in our agent and where it is.
I didn't try to run your `http2-prior-knowledge` test but I'd assume the issue is that we currently do not have any tests for http2 clients or servers so http2 requests not being traced is not unexpected.

**Comment by muhardanz** [comment]
Thanks @laurit. I don't think I will be able to reproduce the OOM errors. These were first observed in a kubernetes cluster useing 1.26.0. At this point I am thinking that its unlikely that the OOM is related to the java agent. Though we did notice the errors started around the time we added in the agent. I think I may have jumped to conclusions correlating the strange trace-id overriding with http2 and the OOM exceptions.

Thanks for the info about the limitations of http 2 and you patience in testing. Much appreciated
